packages/node/src/executors/execute/execute.impl.ts

```diff
@@ -57,68 +71,66 @@ export async function* executeExecutor(
     }
   }
 
-  for await (const event of startBuild(options, context)) {
-    if (!event.success) {
-      logger.error('There was an error with the build. See above.');
-      logger.info(`${event.outfile} was not restarted.`);
+  for await (const event of scheduleBuildAndProcessEvents(
+    startBuild(options, context),
+    processRunner
+  )) {
+    if (event.type === 'build') {
+      if (!event.value.success) {
+        logger.error('There was an error with the build. See above.');
+        logger.info(`${event.value.outfile} was not restarted.`);
+      }
     }
-    await handleBuildEvent(event, options);
-    yield event;
-  }
-}
 
-function runProcess(event: NodeBuildEvent, options: NodeExecuteBuilderOptions) {
-  if (subProcess || !event.success) {
-    return;
+    if (event.type === 'build' || options.emitSubprocessEvents) {
+      yield event.value;
+    }
   }
-
-  subProcess = fork(event.outfile, options.args, {
-    execArgv: getExecArgv(options),
-  });
 }
 
-function getExecArgv(options: NodeExecuteBuilderOptions) {
-  const args = ['-r', 'source-map-support/register', ...options.runtimeArgs];
-
-  if (options.inspect === true) {
-    options.inspect = InspectType.Inspect;
-  }
-
-  if (options.inspect) {
-    args.push(`--${options.inspect}=${options.host}:${options.port}`);
-  }
-
-  return args;
-}
+const nextIterValue = (typedIter: TypedIterator) =>
+  typedIter.type === 'build'
+    ? typedIter.iterator.next().then((result) => ({
+        iterator: typedIter.iterator,
+        type: typedIter.type,
+        ...result,
+      }))
+    : typedIter.iterator.next().then((result) => ({
+        iterator: typedIter.iterator,
+        type: typedIter.type,
+        ...result,
+      }));
+
+async function* scheduleBuildAndProcessEvents(
+  buildIterator: AsyncIterator<NodeBuildEvent>,
+  subProcessRunner: SubProcessRunner
+): AsyncGenerator<ScheduledEvent> {
+  const iteratorMap = new Map<
+    AsyncIterator<SubProcessEvent | NodeBuildEvent>,
+    ReturnType<typeof nextIterValue>
+  >([
+    [buildIterator, nextIterValue({ type: 'build', iterator: buildIterator })],
+  ]);
+
+  while (iteratorMap.size > 0) {
+    const result = await Promise.race(iteratorMap.values());
+
+    if (result.done) {
+      iteratorMap.delete(result.iterator);
+      continue;
+    }
 
-async function handleBuildEvent(
-  event: NodeBuildEvent,
-  options: NodeExecuteBuilderOptions
-) {
-  if ((!event.success || options.watch) && subProcess) {
-    await killProcess();
-  }
-  runProcess(event, options);
-}
+    if (result.type === 'build' && result.value.success) {
+      await subProcessRunner.handleBuildEvent(result.value);
+      iteratorMap.set(
+        subProcessRunner,
+        nextIterValue({ type: 'process', iterator: subProcessRunner })
+      );
+    }
```

Comment:
Did you run into challenges having the `handleBuildEvent` call in the outer function? When you first showed me this function it was more akin to a generic `merge` whereas now it's very impl specific.

e.g.

```ts
for await (const event of scheduleBuildAndProcessEvents(
    startBuild(options, context),
    processRunner
  )) {
    if (event.type === 'build' && event.value.success) {
      processRunner.handleBuildEvent(event.value);
      // ...
```

Reply:
We need to handle these events at the time of scheduling. After our convo opted to move this down into the schedule generator and just `yield * ` to completion